### ISSUE TITLE
`<pre><code>` formatting is broken -- at least on the Planet page

### ISSUE THREAD

**Issue by gasche** [ocaml/v2.ocaml.org#161]
On the planet page, code included from the planet blog posts is not correctly formatted. On [this example](http://ocaml.org/planet/#16902b1ec793bf3116ce09679cdd26aa) from Gagallium, code inside `<pre><code>`  (following the HTML5 recommendation for formatted code) is rendered without the required line breaks.

Other posts on the planet are correctly formatted because they use a server-side renderer that returns HTML code with explicit `<br/>` (but no fixed-width font).

I had a look at the CSS but I'm no expert and I have no idea why the formatting isn't correct.

**Comment by Chris00** [comment]
On Tue, 9 Jul 2013 04:41:53 -0700, gasche wrote:

> On the planet page, code included  from the planet blog posts is not
> correctly  formatted. On  this example  from Gagallium,  code inside
> <pre><code> (following the HTML5  recommendation for formatted code)
> is rendered without the required line breaks.

The problem is _not_ the CSS.  Here is an interactive session (the
Gallium post is the first one at the time of this writing) that shows
the problem origin:

# #require "netclient";;

# #require "rss";;

# let rss = http_get "http://planet.ocaml.org/rss20.xml";;

# String.index rss '\n';;
- : int = 21
  # let r, _ = Rss.channel_of_string rss;;
  # let Some txt = (List.hd r.Rss.ch_items).Rss.item_desc;;
  # String.index txt '\n';;
  Exception: Not_found.

So the problem is that the `Rss` module does not preserve line breaks.

**Comment by dbuenzli** [comment]
Well I don't think it's a problem of the Rss module, IIRC, as far as the RSS spec is concerned the `description` tag is supposed to contain only text (not markup). Gagallium may want to switch to publishing an atom feed instead (which allows to embed xhtml in their descriptions). See also [this comment](https://github.com/ocaml/ocaml.org/issues/123#issuecomment-20552798)

**Comment by amirmc** [comment]
If switching to an Atom feed is not workable (for whatever reason) then perhaps Gagallium can make an Atom feed in addition to the existing RSS.  Then the Atom feed can replace the RSS in Planet.  Would this fix the issue?  If so, it means anyone following the Gagallium feed directly doesn't have to change anything.

**Comment by avsm** [comment]
Cow has an Atom module that is Xmlm based and works fine for several sites (such as openmirage.org)

https://github.com/mirage/ocaml-cow/blob/master/lib/atom.mli

-anil

**Comment by gasche** [comment]
I'm not sure this analysis of the problem is correct; indeed, I do see line breaks (and can find `\n` with a correct adaptation of Cristophe's script) in the output of [http://gallium.inria.fr/blog/index.rss](http://gallium.inria.fr/blog/index.rss). Can you as well? Was there a transformation between the Gagallium feed and the Planet feed?

**Comment by Chris00** [comment]
@gasche May you please specify what you mean by a "correct adaptation" of my script?

I see line breaks in the `rss` string above (so they **are present** as expected at this stage) but I cannot see them anymore in `txt` (and so they do **not** appear in the final HTML).  The only thing in between is the processing by the `Rss` module.

**Comment by gasche** [comment]
I believe one need to `open Http_client.Convenience` before using
`http_get`, and I think my (possibly a bit old) version of the `Rss`
module has `channel_of_string` returns a `Rss.channel` directly
instead of a pair as in your code.

Below is the script I used. In the case of both the Planet RSS feed
and Gagallium's direct feed, I can see the line breaks just fine.

```
#use "topfind";;
#require "netclient";;
#require "rss";;
open Http_client.Convenience;;

let test url =
  let rss = http_get url in
  ignore (try String.index rss '\n' with Not_found -> assert false);
  let r = Rss.channel_of_string rss in
  let our_article = function
    | {Rss.item_guid = Some
        {Rss.guid_name = "http://gallium.inria.fr/blog/gpu_memory_model"; _}; _}
      -> true
    | _ -> false
  in
  let Some txt = (List.find our_article r.Rss.ch_items).Rss.item_desc in
  ignore (String.index txt '\n');
  txt
;;

print_endline (test "http://planet.ocaml.org/rss20.xml");;
print_endline (test "http://gallium.inria.fr/blog/index.rss");;
```

**Comment by Chris00** [comment]
I use the latest OPAM version of RSS: `ocamlrss.2.2.1`.  The mli file says `val channel_of_string : string -> (channel * string list)`.  You seem indeed to use an old version because other modifications to your script were needed to make it run (I also "rewrote" the exception from `String.index` so it is not confused with a possible `Not_found` from `List.find`):

<pre>
#use "topfind";;
#require "netclient";;
#require "rss";;
open Http_client.Convenience;;
let is_gallium url =
  Neturl.string_of_url url
  = "http://gallium.inria.fr/blog/gpu_memory_model"
let test url =
  let rss = http_get url in
  ignore (try String.index rss '\n' with Not_found -> assert false);
  let r, _ = Rss.channel_of_string rss in
  let our_article = function
    | {Rss.item_guid = Some (Rss.Guid_permalink url); _} when is_gallium url
      -> true
    | _ -> false
  in
  let Some txt = (List.find our_article r.Rss.ch_items).Rss.item_desc in
  ignore (try String.index txt '\n' with Not_found -> failwith "No newline");
  txt
;;
print_endline (test "http://planet.ocaml.org/rss20.xml");;
print_endline (test "http://gallium.inria.fr/blog/index.rss");;
</pre>

Running this yields

<pre>
print_endline (test "http://planet.ocaml.org/rss20.xml");;
Exception: (Failure "No newline").
# print_endline (test "http://gallium.inria.fr/blog/index.rss");;
Exception: (Failure "No newline").
</pre>

**Comment by Chris00** [comment]
After more investigations, it is `Xmlm` which removes the newlines...

**Comment by dbuenzli** [comment]
Are you sure the code follows this [tip](http://erratique.ch/software/xmlm/doc/Xmlm#2_Tips).

**Comment by Chris00** [comment]
Thanks to Daniel for the tip.  This [pull request](https://github.com/zoggy/ocamlrss/pull/6) to [OCaml RSS](https://github.com/zoggy/ocamlrss) fixes the issue.

**Comment by Chris00** [comment]
Fixed in commit ae65980dd4b61372021185d9d80c05da8d4abbfc
